lib/Refactoring/SyntacticRenameRangeDetails.cpp

```diff
@@ -366,89 +364,151 @@ RegionType RenameRangeDetailCollector::getSyntacticRenameRegionType(
+  Init,
+  CallAsFunction
+};
 
-  // FIXME: this should only be treated specially for instance methods.
-  bool IsCallAsFunction =
-      Old.base() == "callAsFunction" && Config.IsFunctionLike;
+static SpecialBaseName specialBaseNameFor(const DeclNameViewer &declName) {
+  if (!declName.isFunction()) {
+    return SpecialBaseName::None;
+  }
+  if (declName.base() == "subscript") {
+    // FIXME: Don't handle as special name if it is backticked.
+    return SpecialBaseName::Subscript;
+  } else if (declName.base() == "init") {
+    // FIXME: Don't handle as special name if it is backticked.
+    return SpecialBaseName::Init;
+  } else if (declName.base() == "callAsFunction") {
+    // FIXME: this should only be treated specially for instance methods.
+    return SpecialBaseName::CallAsFunction;
+  } else {
+    return SpecialBaseName::None;
+  }
+}
 
-  bool IsSpecialBase = IsInit || IsSubscript || IsCallAsFunction;
+RegionType RenameRangeDetailCollector::addSyntacticRenameRanges(
+    const ResolvedLoc &resolved, const RenameLoc &config) {
 
-  // Filter out non-semantic special basename locations with no labels.
-  // We've already filtered out those in active code, so these are
-  // any appearance of just 'init', 'subscript', or 'callAsFunction' in
-  // strings, comments, and inactive code.
-  if (IsSpecialBase && (Config.Usage == NameUsage::Unknown &&
-                        Resolved.labelType == LabelRangeType::None))
+  if (!resolved.range.isValid())
     return RegionType::Unmatched;
 
-  if (!Config.IsFunctionLike || !IsSpecialBase) {
-    if (renameBase(Resolved.range, RefactoringRangeKind::BaseName))
-      return RegionType::Mismatch;
+  NameUsage usage = config.Usage;
+
+  auto regionKind = getSyntacticRenameRegionType(resolved);
+
+  SpecialBaseName specialBaseName = specialBaseNameFor(Old);
+
+  if (usage == NameUsage::Unknown) {
+    // Unknown name usage occurs if we don't have an entry in the index that
+    // tells us whether the location is a call, reference or a definition. The
+    // most common reasons why this happens is if the editor is adding syntactic
+    // results (eg. from comments or string literals).
+    //
+    // Determine whether we should include them.
+    if (regionKind == RegionType::ActiveCode) {
+      // If the reference is in active code, we should have had a name usage
+      // from the index. Since we don't, they are likely unrelated symbols that
+      // happen to have the same name. Don't return them as matching ranges.
+      return RegionType::Unmatched;
+    } else if (specialBaseName != SpecialBaseName::None &&
+               resolved.labelType == LabelRangeType::None) {
+      // Filter out non-semantic special basename locations with no labels.
+      // We've already filtered out those in active code, so these are
+      // any appearance of just 'init', 'subscript', or 'callAsFunction' in
+      // strings, comments, and inactive code.
+      return RegionType::Unmatched;
+    }
```

Comment:
Personally I would write these as consecutive `if` statements given the first one returns, but I don't feel too strongly about it

Reply:
Updated